### ISSUE TITLE
pnpm install sometimes fails due to a race condition in the "request" package

### ISSUE THREAD

**Issue by pgonzal** [pnpm/pnpm#953]
### pnpm version:  1.16.0

**request** version: 2.83.0

### Code to reproduce the issue:

See [request issue #2807](https://github.com/request/request/issues/2807) for full details.  We operate an monorepo with around 100 projects and around 200 developers.  Because this issue is a race condition, the repro is difficult to isolate and seems to be influenced by a number of different factors.  We were able to isolate a VM that consistently produces the failure, and can be debugged using the VS Code debugger, but unfortunately we cannot share it publicly because it contains internal code.  (We might be able to share it privately, or provide specific debugging information if requested.)

We have had around 10 different people report this issue so far.  The reports slowed down when we circulated a workaround, which involves running the installation using the [Verdaccio](https://github.com/verdaccio/verdaccio) proxy (which is fairly reliable, but obviously a very poor experience).

### Expected behavior:

The `pnpm install` command should complete successfully.

### Actual behavior:

The `pnpm install` command fails eventually with this error:

```
              events.js:160
                               throw er; // Unhandled 'error' event
                               ^
 
              Error: read ECONNRESET
                             at exports._errnoException (util.js:1018:11)
                             at TLSWrap.onread (net.js:572:26)
```
This issue also happens with NPM, but the error looks like this:

```
npm ERR! code EINTEGRITY
npm ERR! sha1-20E+D1MrUwyeZvYOmVLkQnQ+o1Y= integrity checksum failed when using sha1: wanted sha1-20E+D1MrUwyeZvYOmVLkQnQ+o1Y= but got sha1-ygU0cyjTuZdxezDk6ofWXCGJbqI=. (867808 bytes)
```

### Additional information:

We believe the root cause is a longstanding design flaw in the [request](https://github.com/request/request) library.  I posted a full analysis and details in [request issue #2807](https://github.com/request/request/issues/2807).  It only occurs frequently in large repos (in our case around 1,500 packages including more than 10 packages with relatively large tarballs), however you can find numerous reports of this issue in internet forums, some of which I linked in my analysis.

I also submitted a [PR #2808 for request](https://github.com/request/request/pull/2808) with a workaround that seems to eliminate the race condition for PNPM (but not for NPM).

Unfortunately the **request** project is no longer maintained.  At this point they do not seem to be accepting PRs, and the maintainer only publishes releases for the purpose of bumping dependency versions.

 - `node -v` prints:  6.10.2, and many other versions
 - Windows, OS X, or Linux?: Windows, reproed with many different OS releases

### Recommended Solution

1. The ideal solution would be for **pnpm** to eliminate its dependency on the **npm-registry-client** and **request** libraries.  They are part of the very brittle code base that has driven developers away from NPM.  In particular, we don't believe the race condition can be really solved without a significant reworking of the way they handle redirects, which may break existing clients.  Yarn does not rely on **npm-registry-client** or **request**, so ideally **pnpm** could reuse Yarn's networking layer.

2. Alternatively, **pnpm** could apply our workaround from [request PR #2808](https://github.com/request/request/pull/2808).  This could be done by forking and replacing the **request** library, or by "monkey patching" the code at runtime.

**Comment by pgonzal** [comment]
@nickpape-msft FYI

**Comment by zkochan** [comment]
We are definitely moving away from **npm-registry-client**. I am in a middle of a big rewrite of the package requesting logic (the last related PR https://github.com/pnpm/supi/pull/24, related issue: https://github.com/pnpm/pnpm/issues/559).

After the rewrite, **npm-registry-client** is used by [@pnpm/npm-resolver](https://github.com/pnpm/npm-resolver) and [tarball-fetcher](https://github.com/pnpm/tarball-fetcher). We can use [make-fetch-happen](https://www.npmjs.com/package/make-fetch-happen) instead (with turned off cache).

**Comment by pgonzal** [comment]
Awesome!  Will this also eliminate the dependency on **request**?  I believe the root cause is a design flaw in the **request** library's handling of HTTP redirects.

**Comment by zkochan** [comment]
Yes, it will. make-fetch-happen doesn't have request anywhere in its deps

**Comment by pgonzal** [comment]
Awesome!

**Comment by vjpr** [comment]
@pgonzal I did a long investigation of this same issue. Was happening sporadically on the same 20MB package. I thought it was related to a proxy npm uses but they check it and apparently it wasn't. See https://github.com/npm/registry/issues/202

**Comment by zkochan** [comment]
Published with [pnpm@1.23.2](https://github.com/pnpm/pnpm/releases/tag/v1.23.2)

`request` is still used in node-gyp but that one is not called too frequently, so the error should be gone

**Comment by pgonzal** [comment]
Awesome!  Thanks a ton!
